File: go/examples/print_health_status/print_health_status.go
package main

import (
	"context"
	"fmt"
	"os"

	"github.com/seoulrobotics/horus-sdk/go/horus"
)

// Displays cleanly a health status.
func printHealthStatus(healthStatus *horus.HealthStatus) {
	fmt.Println("---- License status ----")
	fmt.Printf("Level: %d\n", healthStatus.LicenseStatus.LicenseLevel)
	fmt.Printf("Reason: %s\n", healthStatus.LicenseStatus.Reason)

	fmt.Println("Privileges:")
	for privilege, hasPrivilege := range healthStatus.LicenseStatus.Privileges {
		if hasPrivilege {
			fmt.Printf("- %s\n", privilege.String())
		}
	}

	if healthStatus.LicenseStatus.LicenseInfo != nil {
		fmt.Printf("Expiration timestamp: %s\n", healthStatus.LicenseStatus.LicenseInfo.Expiration.String())
		fmt.Printf("Lidar count: %d\n", healthStatus.LicenseStatus.LicenseInfo.LidarCount)

		if len(healthStatus.LicenseStatus.LicenseInfo.AllowedFeatures) > 0 {
			fmt.Println("Allowed features:")
			for _, feature := range healthStatus.LicenseStatus.LicenseInfo.AllowedFeatures {
				fmt.Printf("  - %s\n", feature.String())
			}
		} else {
			fmt.Println("Allowed features: NONE!")
		}
	}
	fmt.Println()

	if !healthStatus.LicenseStatus.HasPrivilege(horus.LicensePrivilegeRead) {
		fmt.Printf("License level does not grant Read privilege. Could not receive sensor and service statuses")
		return
	}

	fmt.Println("---- Sensor statuses ----")
	for _, sensorStatus := range healthStatus.SensorStatuses {
		fmt.Printf("%s:", sensorStatus.LidarId)

		if sensorStatus.IsUnreachable() {
			fmt.Printf(" Unreachable: %s\n\n", sensorStatus.UnreachableReason)
			continue
		}

		fmt.Println()
		fmt.Println("- Statuses:")
		for status, hasStatus := range sensorStatus.Statuses {
			if hasStatus {
				fmt.Printf("  - %s\n", status.String())
			}
		}
		fmt.Printf("- Frequency: %f Hz\n", sensorStatus.MeasuredFrequencyHz)
		fmt.Println()
	}

	fmt.Println("---- Service statuses ----")

	printService := func(service horus.NodeHealthService) {
		var serviceNodes []horus.NodeHealth
		for _, serviceStatus := range healthStatus.ServiceStatuses {
			if serviceStatus.Service == service {
				serviceNodes = append(serviceNodes, serviceStatus)
			}
		}

		fmt.Printf("%s:", service.String())
		if len(serviceNodes) == 0 {
			fmt.Println(" No node found!")
			fmt.Println()
			return
		}
		fmt.Println()
		for _, serviceNode := range serviceNodes {
			fmt.Printf("- %s: %s\n", serviceNode.NodeId, serviceNode.Status.String())
		}
		fmt.Println()
	}

	printService(horus.NodeHealthServiceServiceDetection)
	printService(horus.NodeHealthServiceServiceLidarRunner)
	printService(horus.NodeHealthServiceServiceNotification)
	printService(horus.NodeHealthServiceServicePointAggregator)
	printService(horus.NodeHealthServiceServicePreprocessing)
}

func main() {
	ctx := context.Background()

	// Only create connection to the project manager service
	onlyProjectManagerServices := horus.Services{}.WithProjectManager(*horus.DefaultServices().ProjectManager)

	sdk, err := horus.NewSdk(ctx, horus.SdkOptions{Services: &onlyProjectManagerServices})
	if err != nil {
		fmt.Printf("Error while creating the SDK: %s", err)
		os.Exit(1)
	}

	healthStatus, err := sdk.GetHealthStatus(horus.GetHealthStatusRequest{})
	if err != nil {
		fmt.Printf("Error while getting health status: %s", err)
		os.Exit(1)
	}

	printHealthStatus(healthStatus)
}

File: py/examples/print_health_status.py
import asyncio, horus
from horus.sdk.health import (
    HealthStatus,
    LicensePrivilege,
    NodeHealth,
)
from horus.sdk.version import Version


async def print_health_status() -> None:
    health_status: HealthStatus
    version: Version
    async with horus.Sdk() as sdk:
        health_status = await sdk.get_health_status()
        version = await sdk.get_version()

    print(f"Horus version: {version}\n")

    print("---- License status ----")
    print(f"Licence level: {health_status.license_status.level}")
    print(f"Reason: {health_status.license_status.reason}")

    print("Privileges:")
    for privilege in health_status.license_status.privileges:
        print(f"- {privilege.name}")

    if health_status.license_status.license_info is not None:
        date_str = (
            health_status.license_status.license_info.expiration_datetime.strftime(
                "%Y-%m-%d %H:%M:%S"
            )
        )
        print(f"Expiration date: {date_str}")
        print(
            f"Lidar count: {health_status.license_status.license_info.number_of_lidars}"
        )
        if len(health_status.license_status.license_info.allowed_features) > 0:
            print(f"Allowed features:")
            for feature in health_status.license_status.license_info.allowed_features:
                print(f"  - {feature.name}")
        else:
            print(f"Allowed features: NONE!")

        print()

    if LicensePrivilege.READ not in health_status.license_status.privileges:
        print(
            "Sensor and service statuses not received because license level does not grant the READ privilege."
        )
        return

    print("---- Sensor statuses ----")
    for sensor_status in health_status.sensor_statuses:
        if sensor_status.is_unreachable():
            print(
                f"{sensor_status.lidar_id}: Unreachable! {sensor_status.node_unreachable_error}"
            )
            print()
            continue

        print(f"{sensor_status.lidar_id}:")
        print(f"- Status:")
        for status in sensor_status.status:
            print(f"  - {status.name}")

        print(
            f"- Measured frequency: {sensor_status.measured_frequency_hz}Hz",
        )
        print()

    print("---- Service statuses ----")
    for service in NodeHealth.Service:
        nodes = [
            node for node in health_status.service_statuses if node.service == service
        ]

        if len(nodes) == 0:
            print(f"{service.name}: No node found!\n")
            continue

        print(f"{service.name}")

        for node in nodes:
            print(f"- {node.node_id}: {node.node_status.name}")
        print()


asyncio.run(print_health_status())
